{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f6509d06-3cde-4026-ac42-b76a21962d4a",
   "metadata": {},
   "source": [
    "# Llama-2 7B 4-Bit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3acf0a40-7ab5-467a-baeb-1b420c62c262",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cdsw/.local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "Loading checkpoint shards: 100%|██████████████████| 2/2 [00:50<00:00, 25.21s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running on local URL:  http://127.0.0.1:8091\n",
      "Running on public URL: https://b78fcac35c825a917f.gradio.live\n",
      "\n",
      "This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><iframe src=\"https://b78fcac35c825a917f.gradio.live\" width=\"100%\" height=\"500\" allow=\"autoplay; camera; microphone; clipboard-read; clipboard-write;\" frameborder=\"0\" allowfullscreen></iframe></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gradio app ready\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:>>> 14 LLM Survey Report 2023Not Using Out of the 47 respondents who said they were not using LLMs in production, the most common response (34%) said there was no real use case for the business. The long tail of responses that came up is worth noting. Respondents spoke of data privacy concerns (see Samsung employees using OpenAI). Within the data privacy theme, both data governance and data access were mentioned. Lack of provenance was also cited by respondents who spoke about not knowing what data the model was trained on as being detrimental. Other responses included participants highlighting that the current tooling is not mature enough to make them feel confident, or that they had tried to use LLMs but ultimately gave up as it did not work for their use case.Just as in music, it is not always about the notes you play but rather the notes you don’t play. It is worth pointing out what respondents did not say. Not one response highlighted the lack of budget as a reason for not using LLMs. Respondents know that costs are high; from the survey data this is clear. However, being that projects are so new, budgets may not even be considered yet.It took us less than 3 minutes to get BioGPT to tell us vaccines cause autism\n",
      "\n",
      ">>> 02 LLM Survey Report 2023MLOps Community | Am Sonnenberg 49, Heidenrod 65321, Germany Introduction 04 Who Participated 05 Survey at a glance Limitations of the Analysis Are you using LLM at your organization? Going Deeper than the Common Use Cases 08 Niche Chatbots Customer Support Coding Data Enrichment Product Recommendations Special Tooling 10 Yes we are building special tooling Main Challenges and Concerns 12 Does the high cost actually have and ROI? Trust Not using Open Questions and Opportunities, Looking forward 15 Looking forward About the MLOps Community 19INDEX\n",
      ">>QUESTION<<What were the common use cases identified in the LLM Survey Report ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:Machine Learning Backing up ML workspaces ], \"Effect\": \"Allow\", \"Resource\": \"*\" }, { \"Effect\": \"Allow\", \"Action\": [ \"ec2:CreateTags\", \"ec2:DeleteSnapshot\" ], \"Resource\": \"arn:aws:ec2:*::snapshot/*\" }, { \"Effect\": \"Allow\", \"Action\": [ \"ec2:DeleteSnapshot\", \"ec2:CreateSnapshot\" ], \"Resource\": [ \"arn:aws:ec2:*::snapshot/*\", \"arn:aws:ec2:*:*:volume/*\" ] }, { \"Action\": [ \"ec2:DeleteSnapshot\" ], \"Effect\": \"Allow\", \"Resource\": \"*\", \"Condition\": { \"ForAnyValue:StringEquals\": { \"aws:CalledVia\": [ \"backup.amazonaws.com\" ] } } }, { \"Action\": [ \"tag:GetTagKeys\", \"tag:GetTagValues\", \"tag:GetResources\" ], \"Effect\": \"Allow\", \"Resource\": \"*\" }, { \"Action\": [ \"iam:ListRoles\", \"iam:GetRole\" ], \"Effect\": \"Allow\", \"Resource\": \"*\" }, { \"Effect\": \"Allow\", \"Action\": \"iam:PassRole\", \"Resource\": [ \"arn:aws:iam::*:role/*\" ], \"Condition\": { \"StringLike\": { \"iam:PassedToService\": \"backup.amazonaws.com\" } 14\n",
      "\n",
      "Machine Learning Backing up ML workspaces \"Effect\": \"Allow\", \"Action\": [ \"ec2:CreateTags\" ], \"Resource\": \"arn:aws:ec2:*::snapshot/*\" }, { \"Effect\": \"Allow\", \"Action\": [ \"ec2:DescribeSnapshots\", \"ec2:DescribeVolumes\" ], \"Resource\": \"*\" }, { \"Effect\": \"Allow\", \"Action\": [ \"elasticfilesystem:Restore\", \"elasticfilesystem:CreateFilesystem\", \"elasticfilesystem:DescribeFilesystems\", \"elasticfilesystem:DeleteFilesystem\", \"elasticfilesystem:TagResource\" ], \"Resource\": \"arn:aws:elasticfilesystem:*:*:file-system/*\" }, { \"Effect\": \"Allow\", \"Action\": \"kms:DescribeKey\", \"Resource\": \"*\" }, { \"Effect\": \"Allow\", \"Action\": [ \"kms:Decrypt\", \"kms:Encrypt\", \"kms:GenerateDataKey\", \"kms:ReEncryptTo\", \"kms:ReEncryptFrom\" ], \"Resource\": \"*\", \"Condition\": { \"StringLike\": { \"kms:ViaService\": [ \"ec2.*.amazonaws.com\", \"elasticfilesystem.*.amazonaws.com\" ] } } }, { \"Effect\": \"Allow\", \"Action\": \"kms:CreateGrant\", \"Resource\": \"*\", \"Condition\": { \"Bool\": { \"kms:GrantIsForAWSResource\": \"true\" } } } ] } Save this as cml-restore-policy . 16\n",
      ">>QUESTION<<What are the steps to create an S3 bucket ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:Cloudera Data Engineering Managing a CDE Service 15.Check the box labeled Enable Workload Analytics to automatically send diagnostic information from job execution to Cloudera Workload Manager . 16.Optionally add Tags as needed. Tags are applied to the cloud provider resources associated with the CDE service (including virtual clusters created in that service). For more information about tags, see the cloud provider documentation: Amazon AWS Tagging AWS resources Microsoft Azure Use tags to organize your Azure resources and management hierarchy 17.Default Virtual Cluster selection is enabled by default to create a default virtual cluster after enabling a CDE service. This will help you get a jump start to create your jobs easily, without having to wait to create a CDE virtual cluster as mentioned in Creating virtual clusters , making the onboarding smoother. You can turn this toggle off if you do not wish to use a default virtual cluster. 18.Click Enable. Results The CDE Administration page displays the status of the environment initialization. You can view logs for the environment by clicking on the environment vertical ellipsis menu, and then clicking View Logs . Managing a CDE Service You can view configuration, metrics, and logs of existing CDE services. You can use the Edit option to make the configuration changes dynamically. 1.In the Cloudera Data Platform (CDP) console, click the Data Engineering tile. The CDE Home page displays. 2.Click Administration in the left navigation menu. The Administration page displays. 3.In the Services column, click the Service Details icon for the CDE service you want to manage. 4.On the Configuration tab, you can view details about the service, including the service name and CDP environment. 5.Optional: Click the Edit option to make the configuration changes dynamically which may take a few minutes to update. You can switch between the following tabs to view additional information: Configuration The Configuration tab lists details about the service name, CDP environment, and the CPU and memory capacity. You can modify the CPU and memory capacity dynamically. Charts The Charts tab displays the charts related to CPU Requests, Memory Requests, Jobs, and Nodes. Logs The Logs tab displays the latest log entries for the CDE service. Access The Access tab displays the option to add Amazon Web Services (AWS) user's Amazon Resource Names (ARNs) for the CDE Service. The Access tab also displays the users you have added to the CDE Service which will provide them with the following: •Access to all pods •Access to all secrets including TGTs •Bypass Istio security •Access to tgtgen host keytab with ability to create, delete, and modify users in FreeIPA Diagnostics 6\n",
      "\n",
      "Cloudera Data Engineering Enabling a Cloudera Data Engineering service Enabling a Cloudera Data Engineering service Before you can use the Cloudera Data Engineering (CDE) service, you must enable it on each environment that you want to use CDE on. Before you begin Make sure that you have a working environment for which you want to enable the CDE service. For more information about environments, see Environments . Important : The user interface for CDE 1.17 and above has been updated. The left-hand menu was updated to provide easy access to commonly used pages. The steps below will vary slightly, for example, the Overview page has been replaced with the Home page. You enable a CDE service by clicking Administration on the left-hand menu, then proceed to step 2 listed below. The new home page still displays Virtual Clusters, but now includes quick-access links located at the top for the following categories: Jobs, Resources , and Download & Docs . Note: CDE on Microsoft Azure does not currently support SSD or Spot instances. Procedure 1.In the Cloudera Data Platform (CDP) console, click the Data Engineering tile. The CDE Home page displays. 2.Click Administration in the left navigation menu. The Administration page displays. 3.In the Services column, click the plus icon at the top to enable CDE for an environment. 4.Type the name of the service that you want to enable CDE for. 5.Select the Environment from the drop-down menu. 6.Select the Workload Type . The workload type corresponds to the instance size that will be deployed to run your submitted Spark jobs. When you select a type, the corresponding cloud provider instance size is displayed in the Summary section to the right. a)If you want to use SSD storage, check the box labeled Use SSD instances . In this configuration, SSD storage is used for the workload filesystem, such as the Spark local directory. If your workload requires more space than is available in the instance storage, select a larger instance type with sufficient local storage or select an instance type without SSD, and then configure the EBS volume size. 7.Set the Auto-Scale Range . The range you set here creates an auto scaling group with the specified minimum and maximum number of instances that can be used. The CDE service launches and shuts down instances as needed within this range. The instance size is determined by the Workload Type you selected. 8.If you want to use spot instances, check the box labeled Use Spot instances and select a range of spot instances to request. This creates another auto scaling group of spot instances. Spot instances are requested with similar CPU and memory profiles as the instances selected for the Workload Type . For more information, see Cloudera Data Engineering Spot Instances . 4\n",
      ">>QUESTION<<Describe the minimal setup for AWS Cloud Storage on CDP ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:Machine Learning Provisioning ML Workspaces Provisioning ML Workspaces This topic describes how to provision machine learning (ML) workspaces. Before you begin The first user to access the ML workspace after it is created must have both the MLAdmin role and the EnvironmentAdmin account role assigned. See Configuring User Access to CML and Understanding account roles and resource roles for information about this resource role. Procedure 1.Log in to the CDP web interface. On Public Cloud, log in to https://console.cdp.cloudera.com using your corporate credentials or any other credentials that you received from your CDP administrator. 2.Click ML Workspaces . 3.Click Provision Workspace . 4.Fill out the following fields. •Workspace Name - Give the ML workspace a name. For example, user1_dev . Do not use capital letters in the workspace name. •Select Environment - From the dropdown, select the environment where the ML workspaces must be provisioned. If you do not have any environments available to you in the dropdown, contact your CDP administrator to gain access. Note: You cannot choose an environment when the Environment or associated DataLake and FreeIPA is not in an available or running state. •Existing NFS - (Azure only) Enter the mount path from the environment creation procedure. •NFS Protocol version - (Azure only) Specify the protocol version to use when communicating with the existing NFS server. 4\n",
      "\n",
      "Machine Learning | Contents | iii Contents Provisioning ML Workspaces .................................................................................4 Configuring User Access to CML ...........................................................................7 Granting CDP Users Access to Cloudera Machine Learning Workspaces .........................................................7 Granting Remote Access to ML Workspaces ........................................................8 Accessing ML Workspaces via SOCKS Proxy ......................................................9 Monitoring ML Workspaces .................................................................................10 Suspend and resume ML workspaces ..................................................................11 Backing up ML workspaces ..................................................................................12 Workspace Backup and Restore Prerequisites ...................................................................................................13 Back up an ML workspace ................................................................................................................................17 Restore an ML workspace ..................................................................................................................................18 Restoring to a different environment .................................................................................................................18 Removing ML Workspaces ...................................................................................19 Upgrading ML Workspaces ..................................................................................19 ML Upgrades using Backup/Restore ....................................................................21 Step 1 : Backing up the workspace ....................................................................................................................22 Step 2: Restoring into a new workspace with a different workspace URL/domain endpoint ............................25 Step 3: Delete the backed-up workspace ...........................................................................................................27 Step 4: Restore into a new workspace with same URL/domain endpoint as backed up workspace ..................27 Step 5: Delete the interim restored workspace ..................................................................................................29 Frequently Asked Questions ...............................................................................................................................29 Tagging disks to avoid garbage collection ...........................................................29 Modify Instance Group Type ................................................................................30\n",
      ">>QUESTION<<What are the ports I need to setup in my security group for CDP setup ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:Cloudera Data Engineering Removing a Cloudera Data Engineering service The Diagnostics tab provides option to generate and download diagnostics bundle. Removing a Cloudera Data Engineering service Disabling an existing Cloudera Data Engineering (CDE) service stops all jobs, and deletes all associated virtual clusters and virtual cluster metadata. Do not do this unless you are certain that you no longer need any of these. Disabling CDE does not delete your CDP data. If enabling the CDE service on an environment for the first time fails, you must disable the service before you can try again. In this scenario, there are no clusters or jobs, and you can safely perform this procedure. Before you begin Important : The user interface for CDE 1.17 and above has been updated. The left-hand menu was updated to provide easy access to commonly used pages. The steps below will vary slightly, for example, the Overview page has been replaced with the Home page. You can remove a CDE service by clicking Administration on the left-hand menu, then proceed to step 2 listed below. The new home page still displays Virtual Clusters, but now includes quick-access links located at the top for the following categories: Jobs, Resources , and Download & Docs . Procedure 1.In the Cloudera Data Platform (CDP) console, click the Data Engineering tile and click Overview . 2.In the CDE Services column, click the menu icon for the environment for which you want to disable the CDE service, and then click Disable CDE 3. Warning : Disabling an existing Cloudera Data Engineering (CDE) service stops all jobs, deletes all associated virtual clusters and virtual cluster metadata. Do not do this unless you are certain that you no longer need any of these. Additionally, if you're prompted to perform a Force Disable, in the event that a Disable is not successful, you must perform a manual cleanup of cloud infrastructures such as Security Group, EBS Volume, and S3 Bucket. A manual cleanup is not required for a standard Disable. Confirm that you want to disable CDE by typing the environment name and then clicking Disable. Results The CDE Administration page displays the status of the environment that is being disabled. What to do next If you disabled CDE as a result of a failure to enable CDE for the first time on an environment, resolve any reported issues, and then try again. Limiting Incoming Endpoint Traffic for CDE Services For AWS You can limit incoming endpoint traffic for a CDE service. Before you begin Important : The user interface for CDE 1.17 and above has been updated. The left-hand menu was updated to provide easy access to commonly used pages. The steps below will vary slightly, for example, the Overview page has been replaced with the Home page. The new home page still displays Virtual Clusters, but now includes quick-access links located at the top for the following categories: Jobs, Resources , and Download & Docs . 7\n",
      "\n",
      "Cloudera Data Engineering | Contents | iii Contents Enabling a Cloudera Data Engineering service ....................................................4 Managing a CDE Service .........................................................................................6 Removing a Cloudera Data Engineering service ...................................................7 Limiting Incoming Endpoint Traffic for CDE Services For AWS ......................7\n",
      ">>QUESTION<<What are the steps to remove the Cloudera Data Engineering Service ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:10/5/23, 9:56 PM 352-If I believe that my privacy rights have been violated, when can I submit a complaint | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/352/if-i-believe-that-my-privacy-rights-have-been-violated-when-can-i-submit-a-complaint/index.html 1/3 If I believe that my privacy rights have been violated, when can I submit a complaint? Answer: By law, health care providers (including doctors and hospitals) who engage in certain electronic transactions, health plans, and health care clearinghouses, (collectively, “covered entities”) had until April 14, 2003, to comply with the HIPAA Privacy Rule. (Small health plans had until April 14, 2004, to comply). OCR provides further information on its web site about how to file a complaint. </node/3043> Activities occurring before April 14, 2003, are not subject to the O\u0000ice for Civil Rights (OCR) enforcement actions. A\u0000er that date, a person who believes a covered entity is not complying with a requirement of the Privacy Rule may file with OCR a written complaint, either on paper or electronically. This complaint must be filed within 180 days of when the complainant knew or should have known that the act had occurred. An o\u0000icial website of the United States governmentHere’s how you know Navigate to:</>MENU\n",
      "\n",
      "10/5/23, 9:56 PM 352-If I believe that my privacy rights have been violated, when can I submit a complaint | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/352/if-i-believe-that-my-privacy-rights-have-been-violated-when-can-i-submit-a-complaint/index.html 1/3 If I believe that my privacy rights have been violated, when can I submit a complaint? Answer: By law, health care providers (including doctors and hospitals) who engage in certain electronic transactions, health plans, and health care clearinghouses, (collectively, “covered entities”) had until April 14, 2003, to comply with the HIPAA Privacy Rule. (Small health plans had until April 14, 2004, to comply). OCR provides further information on its web site about how to file a complaint. </node/3043> Activities occurring before April 14, 2003, are not subject to the O\u0000ice for Civil Rights (OCR) enforcement actions. A\u0000er that date, a person who believes a covered entity is not complying with a requirement of the Privacy Rule may file with OCR a written complaint, either on paper or electronically. This complaint must be filed within 180 days of when the complainant knew or should have known that the act had occurred. An o\u0000icial website of the United States governmentHere’s how you know Navigate to:</>MENU\n",
      ">>QUESTION<<Where can I submit a complaint if my HIPAA Privacy Rules have been violated ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:10/5/23, 9:56 PM 352-If I believe that my privacy rights have been violated, when can I submit a complaint | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/352/if-i-believe-that-my-privacy-rights-have-been-violated-when-can-i-submit-a-complaint/index.html 1/3 If I believe that my privacy rights have been violated, when can I submit a complaint? Answer: By law, health care providers (including doctors and hospitals) who engage in certain electronic transactions, health plans, and health care clearinghouses, (collectively, “covered entities”) had until April 14, 2003, to comply with the HIPAA Privacy Rule. (Small health plans had until April 14, 2004, to comply). OCR provides further information on its web site about how to file a complaint. </node/3043> Activities occurring before April 14, 2003, are not subject to the O\u0000ice for Civil Rights (OCR) enforcement actions. A\u0000er that date, a person who believes a covered entity is not complying with a requirement of the Privacy Rule may file with OCR a written complaint, either on paper or electronically. This complaint must be filed within 180 days of when the complainant knew or should have known that the act had occurred. An o\u0000icial website of the United States governmentHere’s how you know Navigate to:</>MENU\n",
      "\n",
      "10/5/23, 9:56 PM 352-If I believe that my privacy rights have been violated, when can I submit a complaint | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/352/if-i-believe-that-my-privacy-rights-have-been-violated-when-can-i-submit-a-complaint/index.html 1/3 If I believe that my privacy rights have been violated, when can I submit a complaint? Answer: By law, health care providers (including doctors and hospitals) who engage in certain electronic transactions, health plans, and health care clearinghouses, (collectively, “covered entities”) had until April 14, 2003, to comply with the HIPAA Privacy Rule. (Small health plans had until April 14, 2004, to comply). OCR provides further information on its web site about how to file a complaint. </node/3043> Activities occurring before April 14, 2003, are not subject to the O\u0000ice for Civil Rights (OCR) enforcement actions. A\u0000er that date, a person who believes a covered entity is not complying with a requirement of the Privacy Rule may file with OCR a written complaint, either on paper or electronically. This complaint must be filed within 180 days of when the complainant knew or should have known that the act had occurred. An o\u0000icial website of the United States governmentHere’s how you know Navigate to:</>MENU\n",
      ">>QUESTION<<Give me the website link for OCR to file my complaint.\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:10/5/23, 9:32 PM 187-What does the HIPAA Privacy Rule do | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/187/what-does-the-hipaa-privacy-rule-do/index.html 2/3For patients – it means being able to make informed choices when seeking care and reimbursement for care based on how personal health information may be used. It enables patients to find out how their information may be used, and about certain disclosures of their information that have been made. It generally limits release of information to the minimum reasonably needed for the purpose of the disclosure. It generally gives patients the right to examine and obtain a copy of their own health records and request corrections. It empowers individuals to control certain uses and disclosures of their health information. Learn more about health information privacy. Created 12/19/02 Content created by O\u0000ice for Civil Rights (OCR) Content last reviewed February 6, 2023 Sign Up for Email Updates Receive the latest updates from the Secretary, Blogs, and News Releases. Sign Up <https://cloud.connect.hhs.gov/sBack to top\n",
      "\n",
      "10/5/23, 9:32 PM 187-What does the HIPAA Privacy Rule do | HHS.gov https://www.hhs.gov/hipaa/for-individuals/faq/187/what-does-the-hipaa-privacy-rule-do/index.html 2/3For patients – it means being able to make informed choices when seeking care and reimbursement for care based on how personal health information may be used. It enables patients to find out how their information may be used, and about certain disclosures of their information that have been made. It generally limits release of information to the minimum reasonably needed for the purpose of the disclosure. It generally gives patients the right to examine and obtain a copy of their own health records and request corrections. It empowers individuals to control certain uses and disclosures of their health information. Learn more about health information privacy. Created 12/19/02 Content created by O\u0000ice for Civil Rights (OCR) Content last reviewed February 6, 2023 Sign Up for Email Updates Receive the latest updates from the Secretary, Blogs, and News Releases. Sign Up <https://cloud.connect.hhs.gov/sBack to top\n",
      ">>QUESTION<<Give a detailed overview of the HIPAA Privacy Rule.\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:179Optimizing exon skipping therapies for DMD T. Y okoT a, W. D uDDY , T. P arTriDge Center for Genetic Medicine, Children’ s National Medical Center, Washington, DC, USAActa Myologica • 2007; XXVI; p. 179-184 Address for correspondence: Toshifumi Yokota, Center for Genetic Medicine Research, Children’s National Medical Center, 111 Michi - gan Ave, NW, Washington, DC, 20010, USA. Fax: +1 202 4766014. E-mail: tyokota@cnmcresearch.orgExon skipping is one of the more promising therapeutic options for Duchenne Muscular Dystrophy (DMD). The idea is to use antisense oligonucleotides to splice out selected exons from the pre-mRNA, at or next to the mutation site, so as to generate a translatable transcript from the mutant dystrophin gene. In principle, the majority of DMD mutations can be rescued by targeting selected exons. Recent developments of antisense oli - gonucleotides (AOs) such as 2`O-methylated antisense oligonu - cleotides (2OMeAOs) or phosphorodiamidate morpholino oli - gomers (morpholinos, PMOs) have made it possible to restore dystrophin expression body-wide in dystrophic mice and dys - trophic dogs by single or multi-exon skipping with no obvious side-effect. Since such treatment would, in many cases, require bespoke design of AOs, it is important to demonstrate treatment of a variety of mutations in dystrophic animals. In-frame dele - tion patterns usually result in a mix of Duchenne and milder Becker Muscular Dystrophy (BMD), but the ratio of Duchenne to Becker varies between patterns, and this provides useful in - formation for selection of the exons that might most profitably be targeted. This review summarizes recent progress in exon skipping therapy and discusses future strategies. Key words: Duchenne/Becker muscular dystrophy, dystrophin, re - vertant fibers, morpholinos Duchenne muscular dystrophy and Becker muscular dystrophy Duchenne muscular dystrophy (DMD) is one of the more common lethal inherited diseases with an incidence of 1 in 3,500 live male births worldwide. The gene mu - tated in DMD and milder Becker muscular dystrophy (BMD), encoding the dystrophin protein (1), is the larg - est known gene, with 79 exons spanning 2.4 Mb of DNA. Disease phenotype varies considerably. In-frame muta - tions are commonly associated with BMD phenotypes ranging from almost asymptomatic to severe DMD-like forms. In contrast, out-of-frame mutations show a strong association with severe DMD phenotypes. This “reading frame rule” holds true for 92% of DMD/BMD cases. For instance, in-frame deletion of exon 16 is not associated with disease (2). Interestingly, the length of deleted area in the gene is not the most important factor in determining clinical phenotype. For example, deletion of exons 45-55, spanning 11 exons, is associated with disease severities ranging from mild to almost asymptomatic dystrophy (3). The middle portion (rod domain) of dystrophin (ex - ons 10-60) has been noted to accommodate large dele - tions without serious phenotypic consequence. The most dramatic example is an in-frame deletion of exons 17-48 (46% of the gene), which was discovered in a very mild case of BMD (4). Dystrophin revertant fibers and exon skipping mechanisms Spontaneous restoration of reading-frame of dys - trophin mRNA by exon-skipping actually occurs in many DMD patients and animal models. Scattered through dys - trophic muscle are occasional foci of “revertant” fibers strongly positive for dystrophin which increase in number during cycles of degeneration/regeneration (5, 6). Ex - pression of dystrophin in these “revertant” fibers results from spontaneous skipping of exons around the primary mutation (7). Such spontaneous exon skipping of disease- causing nonsense mutations occurs in several genes such as the Factor VIII gene in hemophilia, the fibrillin gene in Marfan syndrome and the ornithine d-aminotransferase gene in gyrate atrophy (8, 9). However, in these cases, it has not been established whether the facility to skip exons is focal and clonal as it is in dystrophic muscle. Dystrophin-positive fibers generated by AOs gradually disappear within a month or so after AO injection, but spontaneous “revertant” fibers persist for some years and increase with age (6). This expansion of “revertant” fibers appears to reflect the protective effect of the dys - trophin they contain, but the mechanism whereby each clone consistently and efficiently skips specific exons is not currently understood. Such an understanding might aid the development of long-term AO-induced exon skip - ping therapy. Development of a new generation of antisense oligos Various backbone chemistries of antisense oligonu - cleotide have been tested to overcome the problems of in vivo breakdown of DNA or RNA. Recent explorations of drug-like characteristics of AOs have lead to the develop - ment of oligonucleotides that contain phosphorothioate linkages throughout their length and 2’O-modifications of \n",
      "\n",
      "183Optimizing exon skipping therapies for DMD nundrum is not readily explicable, but it might suggest that there are limits on the size of the rod domain if the hinge domain is lacking, perhaps because hinge domains are needed to give flexibility to rod structures beyond a certain length (34). Clearly, we need a better understanding of these differ - ences if we are to make rational choices of exon-skipping targets that will generate the most functionally effective “quasi-dystrophins”. In this quest for a better understand - ing of the relevant factors, perhaps the most relevant evi - dence is to be found in a detailed study of the genotype/ phenotype correlations in the human DMD and BMD pop - ulations. This requires the assembly of a reliable database founded on the unison of strictly defined clinical criteria with comprehensive and detailed information on the nature of the mutation ideally including the genomic, and the tran - script data, and, where applicable, the exons represented in any partial dystrophin protein produced. It is encouraging to note that a number of databases approaching this level of detail are currently being organized.Acknowledgements Authors thank Drs Akinori Nakamura, Shin-ichi Takeda, and Eric Hoffman for useful discussions. References 1. Hoffman EP, Kunkel LM. Dystrophin: the protein product of the Duchenne muscular dystrophy locus. Cell 1987;51:919-28. 2. Schwartz M, Duno M, Palle AL, Krag T, Vissing J. Deletion of exon 16 of the dystrophin gene is not associated with disease. Hum Mutat 2007;28:205.7 3. Beroud C, Tuffery-Giraud S, Matsuo M, et al. Multiexon skipping leading to an artificial DMD protein lacking amino acids from ex - ons 45 through 55 could rescue up to 63% of patients with Duch - enne muscular dystrophy. Hum Mutat 2007;28:196-202. 4. England SB, Nicholson LVB, Johnson MA, et al. Very mild mus - cular dystrophy associated with the deletion of 46% of dystrophin. Nature 1990;343:180-2. 5. Hoffman EP, Morgan JE, Watkins SC, Partridge TA. Somatic rever - sion/suppression of the mouse mdx phenotype in vivo. J Neurol Sci 1990;99:9-25.Figure 1. Diagrammatic representations of dystrophin structure showing spectrin-type repeats (STRs) for in-frame rod domain deletions within hotspot. Exons 42 to 57 of the dystrophin gene, encoding STRs 16 to 22, are represented at top, with the hinge 3 region labeled. Exon boundaries are represented by vertical lines, left-facing arrows, and right-facing arrows, indicating that the next exon begins at, respectively, the first, second, or third, nucleotide of a codon. STRs are shaded differently and STR labels are in black or white for easier reading. Truncated structures for each deletion pattern are represented, together with a description of the new structure resulting at each join, and the respective numbers of Duchenne and Becker cases. Some truncated structures may contain correctly folded ‘hybrid’ STRs. Deletions having less than ten cases are listed last, without associated diagrams. Numbers were calculated by combining data from selected countries in the Leiden Muscular Dystrophy database, as described in text. \n",
      ">>QUESTION<<How common is Duchenne muscular dystrophy worldwide ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:184T. Yokota et al 6. Yokota T, Lu Q, Morgan JE, et al. Expansion of revertant fibers in dystrophic mdx muscles reflects activity of muscle precursor cells and serves as an index of muscle regeneration. J Cell Sci 2006;119:2679-87. 7. Lu QL, Morris GE, Wilton SD, et al. Massive idiosyncratic exon skipping corrects the nonsense mutation in dystrophic mouse mus - cle and produces functional revertant fibers by clonal expansion. J Cell Biol 2000;148:985-96. 8. Dietz HC, Kendzior RJ Jr. Maintenance of an open reading frame as an additional level of scrutiny during splice site selection. Nat Genet 1994;8:183-8. 9. Naylor J, Brinke A, Hassock S, Green PM, Giannelli F. Charac - teristic mRNA abnormality found in half the patients with severe haemophilia A is due to large DNA inversions. Hum Mol Genet 1993;2:1773-8. 10. Lu QL, Lou F, Bou-Gharios G, et al. Functional amounts of dys - trophin produced by skipping the mutated exon in the mdx dys - trophic mouse. Nat Med 2003;9:1009-14. 11. Lu QL, Rabinowitz A, Chen YC, et al. Systemic delivery of antisense oligoribonucleotide restores dystrophin expression in body-wide skeletal muscles. Proc Natl Acad Sci USA 2005;102:198-203. 12. Summerton J, Weller D. Morpholino antisense oligomers: design, preparation, and properties. Antisense Nucleic Acid Drug Dev 1997;7:187-95. 13. Alter J, Lou F, Rabinowitz A, et al. Systemic delivery of mor - pholino oligonucleotide restores dystrophin expression bodywide and improves dystrophic pathology. Nat Med 2006;12:175-7. 14. Muntoni F, Bushby K, van Ommen G. 128th ENMC Internation - al Workshop on ‘Preclinical optimization and Phase I/II Clinical Trials Using Antisense Oligonucleotides in Duchenne Muscular Dystrophy’. 22-24 October 2004, Naarden, The Netherlands. Neu - romuscul Disord 2005;15:450-7. 15. Adams AM, Harding PL, Iversen PL, et al. Antisense oligonucle - otide induced exon skipping and the dystrophin gene transcript: cocktails and chemistries. BMC Mol Biol 2007;8:57. 16. Aartsma-Rus A, Van Deutekom JC, Fokkema IF, Van Ommen GJ, Den Dunnen JT. Entries in the Leiden Duchenne muscular dystro - phy mutation database: an overview of mutation types and para - doxical cases that confirm the reading-frame rule. Muscle Nerve 2006;34:135-44. 17. Araki E, Nakamura K, Nakao K, et al. Targeted disruption of exon 52 in the mouse dystrophin gene induced muscle degeneration similar to that observed in Duchenne muscular dystrophy. Biochem Biophys Res Commun 1997;238:492-7. 18. Im WB, SF Phelps, EH Copen, et al. Differential expression of dys - trophin isoforms in strains of mdx mice with different mutations. Hum Mol Genet 1996;5:1149-53. 19. Kameya S, Araki E, Katsuki M, et al. Dp260 disrupted mice re -vealed prolonged implicit time of the b-wave in ERG and loss of accumulation of beta-dystroglycan in the outer plexiform layer of the retina. Hum Mol Genet 1997;6:2195-203. 20. Cox GA, Phelps SF, Chapman VM, Chamberlain JS. New mdx mutation disrupts expression of muscle and nonmuscle isoforms of dystrophin. Nat Genet 1993;4:87-93. 21. Haenggi T, Soontornmalai A, Schaub MC, Fritschy JM. The role of utrophin and Dp71 for assembly of different dystrophin-associated protein complexes (DPCs) in the choroid plexus and microvascula - ture of the brain. Neuroscience 2004;129:403-13. 22. Papadopoulos MC, Verkman AS. Aquaporin-4 and brain edema. Pediatr Nephrol 2007;22:778-84. 23. Yokota T, Miyagoe Y , Hosaka Y , et al. Aquaporin-4 is absent at the sarcolemma and at perivascular astrocyte endfeet in a1-syntrophin knockout mice. Proc Japan Acad 2000;76B:22-7. 24. Arechavala-Gomeza V , Graham IR, Popplewell LJ, et al. Compara - tive analysis of antisense oligonucleotide sequences for targeted skipping of exon 51 during dystrophin pre-mRNA splicing in hu - man muscle. Hum Gene Ther 2007;18:798-810. 25. Fall AM, Johnsen R, Honeyman K, et al. Induction of revertant fibres in the mdx mouse using antisense oligonucleotides. Genet Vaccines Ther 2006;4:3-6. 26. Yokota T, Pistilli E, Duddy W, Nagaraju K. Potential of oligonu - cleotide-mediated exon-skipping therapy for Duchenne muscular dystrophy. Expert Opin Biol Ther 2007;7:831-42. 27. Aartsma-Rus A, Janson AA, van Ommen GJ, van Deutekom JC. Antisense-induced exon skipping for duplications in Duchenne muscular dystrophy. BMC Med Genet 2007;8:43-8. 28. Covone AE, Lerone M, Romeo G. Genotype-phenotype correlation and germline mosaicism in DMD/BMD patients with deletions of the dystrophin gene. Hum Genet 1991;87:353-60. 29. Talkop UA, Klaassen T, Piirsoo A, et al. Duchenne and Becker mus - cular dystrophies: an Estonian experience. Brain Dev 1999;21:244- 7. 30. Saadat L, Pittman L, Menhart N. Structural cooperativity in spec - trin type repeats motifs of dystrophin. Biochim Biophys Acta 2006;1764:943-54. 31. Menhart N. Hybrid spectrin type repeats produced by exon-skip - ping in dystrophin. Biochim Biophys Acta 2006;1764:993-9. 32. Amann KJ, Renley BA, Ervasti JM. A cluster of basic repeats in the dystrophin rod domain binds F-actin through an electrostatic interaction. J Biol Chem 1998;273:28419-23. 33. Amann KJ, Guo AW, Ervasti JM. Utrophin lacks the rod domain ac - tin binding activity of dystrophin. J Biol Chem 1999;274:35375-80. 34. Koenig M, Kunkel LM. Detailed analysis of the repeat domain of dystrophin reveals four potential hinge segments that may confer flexibility. J Biol Chem 1990;265:4560-6.\n",
      "\n",
      "Genes are built of exons – pieces that must match their neighbours like a jigsaw. When a gene is read, the exons are assembled together in a process called splicing. When an exon is deleted this can stop the pieces joining and prevent cells producing dystrophin protein. Becker muscular dystrophy – a more mild condition also caused by mutations in the dystrophin gene – is caused by deletions where the pieces can still be joined. This allows cells to produce a smaller protein that retains some function. How does exon skipping work? Exon skipping uses small drugs called antisense oligonucleotides to help cells skip over a specific exon during splicing. This allows cells to join a different set of exons together to produce a protein that is shorter than usual but may have some function. Here, exon 50 has been deleted. By using a drug to skip over exon 51, the cell can join exon 49 and 51 to produce a shorter than normal dystrophin protein. Who can exon skipping treat? Exon skipping as a technology has the potential to treat the majority of those living with Duchenne. However, each exon to be skipped will require a separate drug. This means it will take a long time to develop treatments that can skip all 79 of the exons in the dystrophin gene. Are exon skipping drugs available? In the US, there are two exon skipping drugs approved. Exondys 51 (or eteplirsen) is a drug that skips exon 51 of the dystrophin gene and Vyondys 53 (or golodirsen) skips exon 53. \n",
      ">>QUESTION<<What does splicing mean ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:179Optimizing exon skipping therapies for DMD T. Y okoT a, W. D uDDY , T. P arTriDge Center for Genetic Medicine, Children’ s National Medical Center, Washington, DC, USAActa Myologica • 2007; XXVI; p. 179-184 Address for correspondence: Toshifumi Yokota, Center for Genetic Medicine Research, Children’s National Medical Center, 111 Michi - gan Ave, NW, Washington, DC, 20010, USA. Fax: +1 202 4766014. E-mail: tyokota@cnmcresearch.orgExon skipping is one of the more promising therapeutic options for Duchenne Muscular Dystrophy (DMD). The idea is to use antisense oligonucleotides to splice out selected exons from the pre-mRNA, at or next to the mutation site, so as to generate a translatable transcript from the mutant dystrophin gene. In principle, the majority of DMD mutations can be rescued by targeting selected exons. Recent developments of antisense oli - gonucleotides (AOs) such as 2`O-methylated antisense oligonu - cleotides (2OMeAOs) or phosphorodiamidate morpholino oli - gomers (morpholinos, PMOs) have made it possible to restore dystrophin expression body-wide in dystrophic mice and dys - trophic dogs by single or multi-exon skipping with no obvious side-effect. Since such treatment would, in many cases, require bespoke design of AOs, it is important to demonstrate treatment of a variety of mutations in dystrophic animals. In-frame dele - tion patterns usually result in a mix of Duchenne and milder Becker Muscular Dystrophy (BMD), but the ratio of Duchenne to Becker varies between patterns, and this provides useful in - formation for selection of the exons that might most profitably be targeted. This review summarizes recent progress in exon skipping therapy and discusses future strategies. Key words: Duchenne/Becker muscular dystrophy, dystrophin, re - vertant fibers, morpholinos Duchenne muscular dystrophy and Becker muscular dystrophy Duchenne muscular dystrophy (DMD) is one of the more common lethal inherited diseases with an incidence of 1 in 3,500 live male births worldwide. The gene mu - tated in DMD and milder Becker muscular dystrophy (BMD), encoding the dystrophin protein (1), is the larg - est known gene, with 79 exons spanning 2.4 Mb of DNA. Disease phenotype varies considerably. In-frame muta - tions are commonly associated with BMD phenotypes ranging from almost asymptomatic to severe DMD-like forms. In contrast, out-of-frame mutations show a strong association with severe DMD phenotypes. This “reading frame rule” holds true for 92% of DMD/BMD cases. For instance, in-frame deletion of exon 16 is not associated with disease (2). Interestingly, the length of deleted area in the gene is not the most important factor in determining clinical phenotype. For example, deletion of exons 45-55, spanning 11 exons, is associated with disease severities ranging from mild to almost asymptomatic dystrophy (3). The middle portion (rod domain) of dystrophin (ex - ons 10-60) has been noted to accommodate large dele - tions without serious phenotypic consequence. The most dramatic example is an in-frame deletion of exons 17-48 (46% of the gene), which was discovered in a very mild case of BMD (4). Dystrophin revertant fibers and exon skipping mechanisms Spontaneous restoration of reading-frame of dys - trophin mRNA by exon-skipping actually occurs in many DMD patients and animal models. Scattered through dys - trophic muscle are occasional foci of “revertant” fibers strongly positive for dystrophin which increase in number during cycles of degeneration/regeneration (5, 6). Ex - pression of dystrophin in these “revertant” fibers results from spontaneous skipping of exons around the primary mutation (7). Such spontaneous exon skipping of disease- causing nonsense mutations occurs in several genes such as the Factor VIII gene in hemophilia, the fibrillin gene in Marfan syndrome and the ornithine d-aminotransferase gene in gyrate atrophy (8, 9). However, in these cases, it has not been established whether the facility to skip exons is focal and clonal as it is in dystrophic muscle. Dystrophin-positive fibers generated by AOs gradually disappear within a month or so after AO injection, but spontaneous “revertant” fibers persist for some years and increase with age (6). This expansion of “revertant” fibers appears to reflect the protective effect of the dys - trophin they contain, but the mechanism whereby each clone consistently and efficiently skips specific exons is not currently understood. Such an understanding might aid the development of long-term AO-induced exon skip - ping therapy. Development of a new generation of antisense oligos Various backbone chemistries of antisense oligonu - cleotide have been tested to overcome the problems of in vivo breakdown of DNA or RNA. Recent explorations of drug-like characteristics of AOs have lead to the develop - ment of oligonucleotides that contain phosphorothioate linkages throughout their length and 2’O-modifications of \n",
      "\n",
      "Genes are built of exons – pieces that must match their neighbours like a jigsaw. When a gene is read, the exons are assembled together in a process called splicing. When an exon is deleted this can stop the pieces joining and prevent cells producing dystrophin protein. Becker muscular dystrophy – a more mild condition also caused by mutations in the dystrophin gene – is caused by deletions where the pieces can still be joined. This allows cells to produce a smaller protein that retains some function. How does exon skipping work? Exon skipping uses small drugs called antisense oligonucleotides to help cells skip over a specific exon during splicing. This allows cells to join a different set of exons together to produce a protein that is shorter than usual but may have some function. Here, exon 50 has been deleted. By using a drug to skip over exon 51, the cell can join exon 49 and 51 to produce a shorter than normal dystrophin protein. Who can exon skipping treat? Exon skipping as a technology has the potential to treat the majority of those living with Duchenne. However, each exon to be skipped will require a separate drug. This means it will take a long time to develop treatments that can skip all 79 of the exons in the dystrophin gene. Are exon skipping drugs available? In the US, there are two exon skipping drugs approved. Exondys 51 (or eteplirsen) is a drug that skips exon 51 of the dystrophin gene and Vyondys 53 (or golodirsen) skips exon 53. \n",
      ">>QUESTION<<What is Duchenne muscular dystrophy ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import gradio as gr\n",
    "import shutil\n",
    "import random\n",
    "import time\n",
    "import warnings\n",
    "import bitsandbytes\n",
    "import accelerate\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import textwrap\n",
    "import langchain\n",
    "from langchain.llms import HuggingFacePipeline\n",
    "import torch\n",
    "import transformers\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "from transformers import LlamaTokenizer, LlamaForCausalLM, pipeline\n",
    "### Multi-document retriever\n",
    "from langchain.vectorstores import Chroma, FAISS\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.chains import RetrievalQA, VectorDBQA\n",
    "from langchain.document_loaders import PyPDFLoader\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from InstructorEmbedding import INSTRUCTOR\n",
    "from langchain.embeddings import HuggingFaceInstructEmbeddings\n",
    "import glob\n",
    "from InstructorEmbedding import INSTRUCTOR\n",
    "from langchain.embeddings.sentence_transformer import SentenceTransformerEmbeddings\n",
    "import uuid\n",
    "__import__('pysqlite3')\n",
    "import sys\n",
    "sys.modules['sqlite3'] = sys.modules.pop('pysqlite3')\n",
    "import chromadb\n",
    "from chromadb.config import Settings\n",
    "from langchain.prompts import PromptTemplate\n",
    "langchain.verbose = True\n",
    "import time\n",
    "\n",
    "\n",
    "\n",
    "IP_ADDR=\"18.188.2.246\"\n",
    "chroma = chromadb.HttpClient(host=IP_ADDR, port=8000)\n",
    "\n",
    "access_token = \"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\"\n",
    "hugging_face_model = \"meta-llama/Llama-2-7b-chat-hf\"\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(hugging_face_model, use_auth_token=access_token)\n",
    "\n",
    "llm_model = AutoModelForCausalLM.from_pretrained(hugging_face_model, #meta-llama/Llama-2-13b-chat-hf\n",
    "                                                     load_in_4bit=True,\n",
    "                                                     device_map='auto',\n",
    "                                                     torch_dtype=torch.float16,\n",
    "                                                     low_cpu_mem_usage=True,\n",
    "                                                     use_auth_token=access_token\n",
    "                                                    )\n",
    "max_len = 8192\n",
    "llm_task = \"text-generation\"\n",
    "T = 0.1\n",
    "\n",
    "llm_pipeline = pipeline(\n",
    "    task=llm_task,\n",
    "    model=llm_model, \n",
    "    tokenizer=tokenizer, \n",
    "    max_length=max_len,\n",
    "    temperature=T,\n",
    "    top_p=0.95,\n",
    "    repetition_penalty=1.15\n",
    ")\n",
    "\n",
    "llm = HuggingFacePipeline(pipeline=llm_pipeline)\n",
    "\n",
    "#Uploading Files to target location\n",
    "target = '/home/cdsw/data/'\n",
    "def upload_file(files):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    file_paths = [file.name for file in files]\n",
    "    print(file_paths)\n",
    "    for file in file_paths:\n",
    "        shutil.copy(file, target)\n",
    "    return file_paths\n",
    "\n",
    "\n",
    "#Embedding function which will be used to convert the text to Vector Embeddings\n",
    "embedding_function = SentenceTransformerEmbeddings(model_name=\"all-MiniLM-L6-v2\") #TODO: Find replacement\n",
    "\n",
    "#Defining LangChain's Chroma Client \n",
    "langchain_chroma = Chroma(\n",
    "client=chroma,\n",
    "collection_name=\"default\",\n",
    "embedding_function=embedding_function)\n",
    "\n",
    "#retriever = langchain_chroma.as_retriever(search_kwargs={\"k\": 3, \"search_type\" : \"similarity\"})\n",
    "\n",
    "def create_default_collection():\n",
    "    \"\"\"\n",
    "    Create Default Collection in ChromaDB if no collections exists\n",
    "    \"\"\"\n",
    "    chromadb.create_collection(\"default\")\n",
    "    return \"Created Default Collection\"\n",
    "\n",
    "\n",
    "\n",
    "def collection_lists():\n",
    "    \"\"\"\n",
    "    List All Collections available in ChromaDB\n",
    "    \"\"\"\n",
    "    collection_list = []\n",
    "    chroma_collections = chroma.list_collections()\n",
    "    if chroma_collections == None:\n",
    "        create_default_collection()\n",
    "    else:\n",
    "        for collection in chroma_collections:\n",
    "            collection_list.append(collection.name)\n",
    "    return collection_list\n",
    "\n",
    "collection_list = collection_lists()\n",
    "\n",
    "def embed_documents(collection):\n",
    "    \"\"\"\n",
    "    Given a collection name, this function loads PDF documents from a specified directory, preprocesses their content, \n",
    "    and then embeds the documents into a vector database using Chroma. \n",
    "    After embedding, it deletes the processed PDFs from the directory.\n",
    "    \n",
    "    Args:\n",
    "    - collection (str): Name of the collection to be used in the Chroma vector database.\n",
    "\n",
    "    Returns:\n",
    "    - output (str): Message indicating which documents have been embedded.\n",
    "    \"\"\"\n",
    "    loader = DirectoryLoader(\"/home/cdsw/data/\",\n",
    "                         glob=\"**/*.pdf\",\n",
    "                         loader_cls=PyPDFLoader,\n",
    "                         use_multithreading=True)\n",
    "\n",
    "    documents = loader.load()\n",
    "\n",
    "    for i in range(len(documents)):\n",
    "        documents[i].page_content = documents[i].page_content.replace('\\t', ' ')\\\n",
    "                                                         .replace('\\n', ' ')\\\n",
    "                                                         .replace('       ', ' ')\\\n",
    "                                                         .replace('      ', ' ')\\\n",
    "                                                         .replace('     ', ' ')\\\n",
    "                                                         .replace('    ', ' ')\\\n",
    "                                                         .replace('   ', ' ')\\\n",
    "                                                         .replace('  ', ' ')\n",
    "\n",
    "\n",
    "    langchain_chroma = Chroma(\n",
    "    client=chroma,\n",
    "    collection_name=collection,\n",
    "    embedding_function=embedding_function)    \n",
    "\n",
    "    collection = chroma.get_collection(collection) # Needs to be initialized as LangChain cannot add texts\n",
    "\n",
    "    # Document is chunked per page. Each page will be an entry in the Vector DB\n",
    "    for doc in documents: \n",
    "        collection.add(\n",
    "            ids=[str(uuid.uuid1())], metadatas=doc.metadata, documents=doc.page_content\n",
    "        )\n",
    "\n",
    "    pattern = \"/home/cdsw/data/*.pdf\"\n",
    "    files = glob.glob(pattern)\n",
    "    \n",
    "    output = f\"Documents have been embedded: {files}\"\n",
    "    print(output)\n",
    "    # Deleting Files\n",
    "    for file in files:\n",
    "        os.remove(file)\n",
    "    \n",
    "    return output\n",
    "\n",
    "\n",
    "##### Experimentatal Code ##### \n",
    "def set_retriver(collection_name):\n",
    "    langchain_chroma = Chroma(\n",
    "    client=chroma,\n",
    "    collection_name= collection_name,\n",
    "    embedding_function=embedding_function)\n",
    "    \n",
    "    retriever = langchain_chroma.as_retriever(search_kwargs={\"k\": 2, \"search_type\" : \"similarity\"})\n",
    "    return retriever\n",
    "\n",
    "\n",
    "# Prompt Template for Langchain\n",
    "template = \"\"\"You are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
    "Context:{context}\n",
    ">>QUESTION<<{question}\n",
    ">>ANSWER<<\"\"\"\n",
    "QA_CHAIN_PROMPT = PromptTemplate.from_template(template)\n",
    "\n",
    "  \n",
    "\n",
    "def chain(query, retriever):\n",
    "    \"\"\"\n",
    "    Executes a retrieval-based question-answering chain with specified query and retriever.\n",
    "\n",
    "    Args:\n",
    "    - query (str): The query/question to be answered.\n",
    "    - retriever (Retriever): The retriever object responsible for fetching relevant documents.\n",
    "\n",
    "    Returns:\n",
    "    - dict: Response from the RetrievalQA.\n",
    "    \"\"\"\n",
    "    qa_chain = RetrievalQA.from_chain_type(llm=llm, \n",
    "                                       chain_type=\"stuff\", \n",
    "                                       retriever=set_retriver(retriever), \n",
    "                                       return_source_documents=True,\n",
    "                                       chain_type_kwargs={\"prompt\": QA_CHAIN_PROMPT},\n",
    "                                       verbose=True, tags=[\"llama-2-7b\", \"4-bit\"]\n",
    "                                       )\n",
    "    return qa_chain(query)\n",
    "\n",
    "def add_text(history, text):\n",
    "    \"\"\"\n",
    "    Adds the user's text input to the conversation history.\n",
    "\n",
    "    Args:\n",
    "    - history (list): The existing conversation history.\n",
    "    - text (str): The user's input text.\n",
    "\n",
    "    Returns:\n",
    "    - list: Updated history with the user's input.\n",
    "    - str: Empty string (reserved for future use).\n",
    "    \"\"\"\n",
    "    history = history + [(text, None)]\n",
    "    return history, \"\"\n",
    "\n",
    "def bot(history, collection):\n",
    "    \"\"\"\n",
    "    Generates a response using a Language Model and updates the conversation history.\n",
    "\n",
    "    Args:\n",
    "    - history (list): The existing conversation history.\n",
    "    - collection (str): The name of the collection used for document retrieval.\n",
    "\n",
    "    Returns:\n",
    "    - list: Updated conversation history including the bot's response.\n",
    "    \"\"\"\n",
    "    response = llm_ans(history[-1][0], collection)\n",
    "    history[-1][1] = response\n",
    "    return history\n",
    "\n",
    "def wrap_text_preserve_newlines(text, width=110):\n",
    "    \"\"\"\n",
    "    Wraps the text while preserving newlines to fit within a specified width.\n",
    "\n",
    "    Args:\n",
    "    - text (str): The input text.\n",
    "    - width (int): The maximum width of the text.\n",
    "\n",
    "    Returns:\n",
    "    - str: Wrapped text with newlines preserved.\n",
    "    \"\"\"\n",
    "    # Split the input text into lines based on newline characters\n",
    "    lines = text.split('\\n')\n",
    "\n",
    "    # Wrap each line individually\n",
    "    wrapped_lines = [textwrap.fill(line, width=width) for line in lines]\n",
    "\n",
    "    # Join the wrapped lines back together using newline characters\n",
    "    wrapped_text = '\\n'.join(wrapped_lines)\n",
    "\n",
    "    return wrapped_text\n",
    "\n",
    "def process_llm_response(llm_response):\n",
    "    \"\"\"\n",
    "    Processes the Language Model's response by wrapping the text and printing the source documents.\n",
    "\n",
    "    Args:\n",
    "    - llm_response (dict): The response from the Language Model.\n",
    "\n",
    "    Returns:\n",
    "    - str: The wrapped text.\n",
    "    \"\"\"\n",
    "    result = wrap_text_preserve_newlines(llm_response['result'])\n",
    "    print('\\n\\nSources:')\n",
    "    for source in llm_response[\"source_documents\"]:\n",
    "        print(source.metadata['source'])\n",
    "    return result    \n",
    "\n",
    " \n",
    "\n",
    "def llm_ans(query, collection):\n",
    "    \"\"\"\n",
    "    Gets the answer from the Language Model including relevant source files.\n",
    "\n",
    "    Args:\n",
    "    - query (str): The query/question to be answered.\n",
    "    - collection (str): The name of the collection used for document retrieval.\n",
    "\n",
    "    Returns:\n",
    "    - str: The answer along with relevant source files.\n",
    "    \"\"\"\n",
    "    start = time.time()\n",
    "    llm_response = chain(query, collection)\n",
    "    end = time.time()\n",
    "    elapsed_time = end - start\n",
    "\n",
    "    \n",
    "    # print(llm_response['result'])\n",
    "    sources = []\n",
    "    for source in llm_response[\"source_documents\"]:\n",
    "        source_file = source.metadata['source']\n",
    "        source_file = source_file.replace(\"/home/cdsw/data/\", \"\")\n",
    "        sources.append(source_file)\n",
    "    source_files = \"\\n\".join(sources) \n",
    "    ans = llm_response['result'] + \"\\n \\n Relevant Sources: \\n\" + source_files + \"\\n \\n Elapsed Time: \" + str(round(elapsed_time,2)) + \" seconds\"\n",
    "    return ans\n",
    "\n",
    "def reset_state():\n",
    "    \"\"\"\n",
    "    Resets the Gradio UI\n",
    "    \"\"\"\n",
    "    return [], [], None\n",
    "\n",
    "#Gradio UI Code Block\n",
    "\n",
    "with gr.Blocks() as demo:\n",
    "    with gr.Tab(\"FileGPT\"):\n",
    "        chatbot = gr.Chatbot([], elem_id=\"chatbot\").style(height=650)\n",
    "        with gr.Row():\n",
    "            with gr.Column(scale=4):\n",
    "                with gr.Column(scale=12):\n",
    "                    user_input = gr.Textbox(show_label=False, placeholder=\"Input...\", lines=10).style(\n",
    "                        container=False)\n",
    "                with gr.Column(min_width=32, scale=1):\n",
    "                    submitBtn = gr.Button(\"Submit\", variant=\"primary\")\n",
    "            with gr.Column(scale=1):\n",
    "                collection_dropdown = gr.Dropdown(\n",
    "                    collection_list, label=\"Chroma Collections\", info=\"Choose a Collection to Query\",\n",
    "                    value = \"default\", max_choices=1\n",
    "                )\n",
    "                emptyBtn = gr.Button(\"Clear History\")\n",
    "        user_input.submit(add_text, [chatbot, user_input], [chatbot, user_input]).then(bot, [chatbot, collection_dropdown], chatbot)\n",
    "        submitBtn.click(add_text, [chatbot, user_input], [chatbot, user_input]).then(bot, [chatbot, collection_dropdown], chatbot)\n",
    "        history = gr.State([])\n",
    "        past_key_values = gr.State(None)\n",
    "        emptyBtn.click(reset_state, outputs=[chatbot, history, past_key_values], show_progress=True)\n",
    "        \n",
    "\n",
    "    with gr.Tab(\"Upload File\"):\n",
    "        with gr.Row():\n",
    "            title=\"Falcon 40B\",\n",
    "            with gr.Column(scale=4):\n",
    "                file_output = gr.File()\n",
    "                upload_button = gr.UploadButton(\"Click to Upload a File\", file_types=[\".pdf\",\".csv\",\".doc\"], file_count=\"multiple\")\n",
    "                upload_button.upload(upload_file, upload_button, file_output)\n",
    "            with gr.Column(scale=1):\n",
    "                embed_dropdown = gr.Dropdown(\n",
    "                    collection_list, label=\"Chroma Collections\", info=\"Choose a Collection to Query\", \n",
    "                    value = \"default\", max_choices=1\n",
    "                )\n",
    "                embed_button = gr.Button(\"Embed Document\", variant=\"primary\")\n",
    "                txt_3 = gr.Textbox(value=\"\", label=\"Output\")\n",
    "                \n",
    "                \n",
    "    embed_button.click(embed_documents, embed_dropdown, show_progress=True, outputs=[txt_3])\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "demo.queue()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    demo.launch(share=True,\n",
    "                enable_queue=True,\n",
    "                show_error=True,\n",
    "                server_name='127.0.0.1',\n",
    "                server_port=int(\"8091\")) \n",
    "\n",
    "    print(\"Gradio app ready\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f429ba5-a677-41d4-a48b-5b91a1a39d5d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
