{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2fcdbf49-97f6-4b98-8bba-faa96b72cc3b",
   "metadata": {},
   "source": [
    "# Zephyr 7B 4 Bit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e0921975-4292-4e03-84a0-8c4994a70f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cdsw/.local/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mNo Traceloop API key provided, auto-creating a dashboard on Traceloop\n",
      "\u001b[33mTRACELOOP_API_KEY= 7f0f500a8fd2f7f3abdb142c823bc196943a5577b8e791bc6676aeaa1c08b7ad40c1ee8116e3de45f60485efb369fcae\n",
      "\u001b[32m\n",
      "Go to https://app.traceloop.com/trace?skt=vZ8wgkB51N1KLqSymM7vab to see a live dashboard\n",
      "\n",
      "\u001b[32mTraceloop exporting traces to https://api.traceloop.com authenticating with bearer token\n",
      "\u001b[39m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading checkpoint shards: 100%|██████████████████| 8/8 [00:50<00:00,  6.32s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running on local URL:  http://127.0.0.1:8091\n",
      "Running on public URL: https://0e1622d7ffc31300e6.gradio.live\n",
      "\n",
      "This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><iframe src=\"https://0e1622d7ffc31300e6.gradio.live\" width=\"100%\" height=\"500\" allow=\"autoplay; camera; microphone; clipboard-read; clipboard-write;\" frameborder=\"0\" allowfullscreen></iframe></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gradio app ready\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:184T. Yokota et al 6. Yokota T, Lu Q, Morgan JE, et al. Expansion of revertant fibers in dystrophic mdx muscles reflects activity of muscle precursor cells and serves as an index of muscle regeneration. J Cell Sci 2006;119:2679-87. 7. Lu QL, Morris GE, Wilton SD, et al. Massive idiosyncratic exon skipping corrects the nonsense mutation in dystrophic mouse mus - cle and produces functional revertant fibers by clonal expansion. J Cell Biol 2000;148:985-96. 8. Dietz HC, Kendzior RJ Jr. Maintenance of an open reading frame as an additional level of scrutiny during splice site selection. Nat Genet 1994;8:183-8. 9. Naylor J, Brinke A, Hassock S, Green PM, Giannelli F. Charac - teristic mRNA abnormality found in half the patients with severe haemophilia A is due to large DNA inversions. Hum Mol Genet 1993;2:1773-8. 10. Lu QL, Lou F, Bou-Gharios G, et al. Functional amounts of dys - trophin produced by skipping the mutated exon in the mdx dys - trophic mouse. Nat Med 2003;9:1009-14. 11. Lu QL, Rabinowitz A, Chen YC, et al. Systemic delivery of antisense oligoribonucleotide restores dystrophin expression in body-wide skeletal muscles. Proc Natl Acad Sci USA 2005;102:198-203. 12. Summerton J, Weller D. Morpholino antisense oligomers: design, preparation, and properties. Antisense Nucleic Acid Drug Dev 1997;7:187-95. 13. Alter J, Lou F, Rabinowitz A, et al. Systemic delivery of mor - pholino oligonucleotide restores dystrophin expression bodywide and improves dystrophic pathology. Nat Med 2006;12:175-7. 14. Muntoni F, Bushby K, van Ommen G. 128th ENMC Internation - al Workshop on ‘Preclinical optimization and Phase I/II Clinical Trials Using Antisense Oligonucleotides in Duchenne Muscular Dystrophy’. 22-24 October 2004, Naarden, The Netherlands. Neu - romuscul Disord 2005;15:450-7. 15. Adams AM, Harding PL, Iversen PL, et al. Antisense oligonucle - otide induced exon skipping and the dystrophin gene transcript: cocktails and chemistries. BMC Mol Biol 2007;8:57. 16. Aartsma-Rus A, Van Deutekom JC, Fokkema IF, Van Ommen GJ, Den Dunnen JT. Entries in the Leiden Duchenne muscular dystro - phy mutation database: an overview of mutation types and para - doxical cases that confirm the reading-frame rule. Muscle Nerve 2006;34:135-44. 17. Araki E, Nakamura K, Nakao K, et al. Targeted disruption of exon 52 in the mouse dystrophin gene induced muscle degeneration similar to that observed in Duchenne muscular dystrophy. Biochem Biophys Res Commun 1997;238:492-7. 18. Im WB, SF Phelps, EH Copen, et al. Differential expression of dys - trophin isoforms in strains of mdx mice with different mutations. Hum Mol Genet 1996;5:1149-53. 19. Kameya S, Araki E, Katsuki M, et al. Dp260 disrupted mice re -vealed prolonged implicit time of the b-wave in ERG and loss of accumulation of beta-dystroglycan in the outer plexiform layer of the retina. Hum Mol Genet 1997;6:2195-203. 20. Cox GA, Phelps SF, Chapman VM, Chamberlain JS. New mdx mutation disrupts expression of muscle and nonmuscle isoforms of dystrophin. Nat Genet 1993;4:87-93. 21. Haenggi T, Soontornmalai A, Schaub MC, Fritschy JM. The role of utrophin and Dp71 for assembly of different dystrophin-associated protein complexes (DPCs) in the choroid plexus and microvascula - ture of the brain. Neuroscience 2004;129:403-13. 22. Papadopoulos MC, Verkman AS. Aquaporin-4 and brain edema. Pediatr Nephrol 2007;22:778-84. 23. Yokota T, Miyagoe Y , Hosaka Y , et al. Aquaporin-4 is absent at the sarcolemma and at perivascular astrocyte endfeet in a1-syntrophin knockout mice. Proc Japan Acad 2000;76B:22-7. 24. Arechavala-Gomeza V , Graham IR, Popplewell LJ, et al. Compara - tive analysis of antisense oligonucleotide sequences for targeted skipping of exon 51 during dystrophin pre-mRNA splicing in hu - man muscle. Hum Gene Ther 2007;18:798-810. 25. Fall AM, Johnsen R, Honeyman K, et al. Induction of revertant fibres in the mdx mouse using antisense oligonucleotides. Genet Vaccines Ther 2006;4:3-6. 26. Yokota T, Pistilli E, Duddy W, Nagaraju K. Potential of oligonu - cleotide-mediated exon-skipping therapy for Duchenne muscular dystrophy. Expert Opin Biol Ther 2007;7:831-42. 27. Aartsma-Rus A, Janson AA, van Ommen GJ, van Deutekom JC. Antisense-induced exon skipping for duplications in Duchenne muscular dystrophy. BMC Med Genet 2007;8:43-8. 28. Covone AE, Lerone M, Romeo G. Genotype-phenotype correlation and germline mosaicism in DMD/BMD patients with deletions of the dystrophin gene. Hum Genet 1991;87:353-60. 29. Talkop UA, Klaassen T, Piirsoo A, et al. Duchenne and Becker mus - cular dystrophies: an Estonian experience. Brain Dev 1999;21:244- 7. 30. Saadat L, Pittman L, Menhart N. Structural cooperativity in spec - trin type repeats motifs of dystrophin. Biochim Biophys Acta 2006;1764:943-54. 31. Menhart N. Hybrid spectrin type repeats produced by exon-skip - ping in dystrophin. Biochim Biophys Acta 2006;1764:993-9. 32. Amann KJ, Renley BA, Ervasti JM. A cluster of basic repeats in the dystrophin rod domain binds F-actin through an electrostatic interaction. J Biol Chem 1998;273:28419-23. 33. Amann KJ, Guo AW, Ervasti JM. Utrophin lacks the rod domain ac - tin binding activity of dystrophin. J Biol Chem 1999;274:35375-80. 34. Koenig M, Kunkel LM. Detailed analysis of the repeat domain of dystrophin reveals four potential hinge segments that may confer flexibility. J Biol Chem 1990;265:4560-6.\n",
      "\n",
      "Genes are built of exons – pieces that must match their neighbours like a jigsaw. When a gene is read, the exons are assembled together in a process called splicing. When an exon is deleted this can stop the pieces joining and prevent cells producing dystrophin protein. Becker muscular dystrophy – a more mild condition also caused by mutations in the dystrophin gene – is caused by deletions where the pieces can still be joined. This allows cells to produce a smaller protein that retains some function. How does exon skipping work? Exon skipping uses small drugs called antisense oligonucleotides to help cells skip over a specific exon during splicing. This allows cells to join a different set of exons together to produce a protein that is shorter than usual but may have some function. Here, exon 50 has been deleted. By using a drug to skip over exon 51, the cell can join exon 49 and 51 to produce a shorter than normal dystrophin protein. Who can exon skipping treat? Exon skipping as a technology has the potential to treat the majority of those living with Duchenne. However, each exon to be skipped will require a separate drug. This means it will take a long time to develop treatments that can skip all 79 of the exons in the dystrophin gene. Are exon skipping drugs available? In the US, there are two exon skipping drugs approved. Exondys 51 (or eteplirsen) is a drug that skips exon 51 of the dystrophin gene and Vyondys 53 (or golodirsen) skips exon 53. \n",
      ">>QUESTION<<What does splicing mean ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new RetrievalQA chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mYou are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
      "Context:179Optimizing exon skipping therapies for DMD T. Y okoT a, W. D uDDY , T. P arTriDge Center for Genetic Medicine, Children’ s National Medical Center, Washington, DC, USAActa Myologica • 2007; XXVI; p. 179-184 Address for correspondence: Toshifumi Yokota, Center for Genetic Medicine Research, Children’s National Medical Center, 111 Michi - gan Ave, NW, Washington, DC, 20010, USA. Fax: +1 202 4766014. E-mail: tyokota@cnmcresearch.orgExon skipping is one of the more promising therapeutic options for Duchenne Muscular Dystrophy (DMD). The idea is to use antisense oligonucleotides to splice out selected exons from the pre-mRNA, at or next to the mutation site, so as to generate a translatable transcript from the mutant dystrophin gene. In principle, the majority of DMD mutations can be rescued by targeting selected exons. Recent developments of antisense oli - gonucleotides (AOs) such as 2`O-methylated antisense oligonu - cleotides (2OMeAOs) or phosphorodiamidate morpholino oli - gomers (morpholinos, PMOs) have made it possible to restore dystrophin expression body-wide in dystrophic mice and dys - trophic dogs by single or multi-exon skipping with no obvious side-effect. Since such treatment would, in many cases, require bespoke design of AOs, it is important to demonstrate treatment of a variety of mutations in dystrophic animals. In-frame dele - tion patterns usually result in a mix of Duchenne and milder Becker Muscular Dystrophy (BMD), but the ratio of Duchenne to Becker varies between patterns, and this provides useful in - formation for selection of the exons that might most profitably be targeted. This review summarizes recent progress in exon skipping therapy and discusses future strategies. Key words: Duchenne/Becker muscular dystrophy, dystrophin, re - vertant fibers, morpholinos Duchenne muscular dystrophy and Becker muscular dystrophy Duchenne muscular dystrophy (DMD) is one of the more common lethal inherited diseases with an incidence of 1 in 3,500 live male births worldwide. The gene mu - tated in DMD and milder Becker muscular dystrophy (BMD), encoding the dystrophin protein (1), is the larg - est known gene, with 79 exons spanning 2.4 Mb of DNA. Disease phenotype varies considerably. In-frame muta - tions are commonly associated with BMD phenotypes ranging from almost asymptomatic to severe DMD-like forms. In contrast, out-of-frame mutations show a strong association with severe DMD phenotypes. This “reading frame rule” holds true for 92% of DMD/BMD cases. For instance, in-frame deletion of exon 16 is not associated with disease (2). Interestingly, the length of deleted area in the gene is not the most important factor in determining clinical phenotype. For example, deletion of exons 45-55, spanning 11 exons, is associated with disease severities ranging from mild to almost asymptomatic dystrophy (3). The middle portion (rod domain) of dystrophin (ex - ons 10-60) has been noted to accommodate large dele - tions without serious phenotypic consequence. The most dramatic example is an in-frame deletion of exons 17-48 (46% of the gene), which was discovered in a very mild case of BMD (4). Dystrophin revertant fibers and exon skipping mechanisms Spontaneous restoration of reading-frame of dys - trophin mRNA by exon-skipping actually occurs in many DMD patients and animal models. Scattered through dys - trophic muscle are occasional foci of “revertant” fibers strongly positive for dystrophin which increase in number during cycles of degeneration/regeneration (5, 6). Ex - pression of dystrophin in these “revertant” fibers results from spontaneous skipping of exons around the primary mutation (7). Such spontaneous exon skipping of disease- causing nonsense mutations occurs in several genes such as the Factor VIII gene in hemophilia, the fibrillin gene in Marfan syndrome and the ornithine d-aminotransferase gene in gyrate atrophy (8, 9). However, in these cases, it has not been established whether the facility to skip exons is focal and clonal as it is in dystrophic muscle. Dystrophin-positive fibers generated by AOs gradually disappear within a month or so after AO injection, but spontaneous “revertant” fibers persist for some years and increase with age (6). This expansion of “revertant” fibers appears to reflect the protective effect of the dys - trophin they contain, but the mechanism whereby each clone consistently and efficiently skips specific exons is not currently understood. Such an understanding might aid the development of long-term AO-induced exon skip - ping therapy. Development of a new generation of antisense oligos Various backbone chemistries of antisense oligonu - cleotide have been tested to overcome the problems of in vivo breakdown of DNA or RNA. Recent explorations of drug-like characteristics of AOs have lead to the develop - ment of oligonucleotides that contain phosphorothioate linkages throughout their length and 2’O-modifications of \n",
      "\n",
      "Genes are built of exons – pieces that must match their neighbours like a jigsaw. When a gene is read, the exons are assembled together in a process called splicing. When an exon is deleted this can stop the pieces joining and prevent cells producing dystrophin protein. Becker muscular dystrophy – a more mild condition also caused by mutations in the dystrophin gene – is caused by deletions where the pieces can still be joined. This allows cells to produce a smaller protein that retains some function. How does exon skipping work? Exon skipping uses small drugs called antisense oligonucleotides to help cells skip over a specific exon during splicing. This allows cells to join a different set of exons together to produce a protein that is shorter than usual but may have some function. Here, exon 50 has been deleted. By using a drug to skip over exon 51, the cell can join exon 49 and 51 to produce a shorter than normal dystrophin protein. Who can exon skipping treat? Exon skipping as a technology has the potential to treat the majority of those living with Duchenne. However, each exon to be skipped will require a separate drug. This means it will take a long time to develop treatments that can skip all 79 of the exons in the dystrophin gene. Are exon skipping drugs available? In the US, there are two exon skipping drugs approved. Exondys 51 (or eteplirsen) is a drug that skips exon 51 of the dystrophin gene and Vyondys 53 (or golodirsen) skips exon 53. \n",
      ">>QUESTION<<What is Duchenne muscular dystrophy ?\n",
      ">>ANSWER<<\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import gradio as gr\n",
    "import shutil\n",
    "import random\n",
    "import time\n",
    "import warnings\n",
    "import bitsandbytes\n",
    "import accelerate\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import textwrap\n",
    "import langchain\n",
    "from langchain.llms import HuggingFacePipeline\n",
    "import torch\n",
    "import transformers\n",
    "from transformers import AutoTokenizer, AutoModelForCausalLM\n",
    "from transformers import LlamaTokenizer, LlamaForCausalLM, pipeline\n",
    "### Multi-document retriever\n",
    "from langchain.vectorstores import Chroma, FAISS\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.chains import RetrievalQA, VectorDBQA\n",
    "from langchain.document_loaders import PyPDFLoader\n",
    "from langchain.document_loaders import DirectoryLoader\n",
    "from InstructorEmbedding import INSTRUCTOR\n",
    "from langchain.embeddings import HuggingFaceInstructEmbeddings\n",
    "import glob\n",
    "from InstructorEmbedding import INSTRUCTOR\n",
    "from langchain.embeddings.sentence_transformer import SentenceTransformerEmbeddings\n",
    "import uuid\n",
    "__import__('pysqlite3')\n",
    "import sys\n",
    "sys.modules['sqlite3'] = sys.modules.pop('pysqlite3')\n",
    "import chromadb\n",
    "from chromadb.config import Settings\n",
    "from langchain.prompts import PromptTemplate\n",
    "langchain.verbose = True\n",
    "import time\n",
    "\n",
    "from traceloop.sdk import Traceloop\n",
    "Traceloop.init(app_name=\"zephyr-7B\", disable_batch=True)\n",
    "\n",
    "\n",
    "IP_ADDR=\"18.188.2.246\"\n",
    "chroma = chromadb.HttpClient(host=IP_ADDR, port=8000)\n",
    "\n",
    "access_token = \"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\"\n",
    "hugging_face_model = \"HuggingFaceH4/zephyr-7b-alpha\"\n",
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(hugging_face_model, use_auth_token=access_token)\n",
    "\n",
    "llm_model = AutoModelForCausalLM.from_pretrained(hugging_face_model, #meta-llama/Llama-2-13b-chat-hf\n",
    "                                                     load_in_4bit=True,\n",
    "                                                     device_map='auto',\n",
    "                                                     torch_dtype=torch.float16,\n",
    "                                                     low_cpu_mem_usage=True,\n",
    "                                                     use_auth_token=access_token\n",
    "                                                    )\n",
    "max_len = 8192\n",
    "llm_task = \"text-generation\"\n",
    "T = 0.1\n",
    "\n",
    "llm_pipeline = pipeline(\n",
    "    task=llm_task,\n",
    "    model=llm_model, \n",
    "    tokenizer=tokenizer, \n",
    "    max_length=max_len,\n",
    "    temperature=T,\n",
    "    top_p=0.95,\n",
    "    repetition_penalty=1.15\n",
    ")\n",
    "\n",
    "llm = HuggingFacePipeline(pipeline=llm_pipeline)\n",
    "\n",
    "#Uploading Files to target location\n",
    "target = '/home/cdsw/data/'\n",
    "def upload_file(files):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    file_paths = [file.name for file in files]\n",
    "    print(file_paths)\n",
    "    for file in file_paths:\n",
    "        shutil.copy(file, target)\n",
    "    return file_paths\n",
    "\n",
    "\n",
    "#Embedding function which will be used to convert the text to Vector Embeddings\n",
    "embedding_function = SentenceTransformerEmbeddings(model_name=\"all-MiniLM-L6-v2\") #TODO: Find replacement\n",
    "\n",
    "#Defining LangChain's Chroma Client \n",
    "langchain_chroma = Chroma(\n",
    "client=chroma,\n",
    "collection_name=\"default\",\n",
    "embedding_function=embedding_function)\n",
    "\n",
    "#retriever = langchain_chroma.as_retriever(search_kwargs={\"k\": 3, \"search_type\" : \"similarity\"})\n",
    "\n",
    "def create_default_collection():\n",
    "    \"\"\"\n",
    "    Create Default Collection in ChromaDB if no collections exists\n",
    "    \"\"\"\n",
    "    chromadb.create_collection(\"default\")\n",
    "    return \"Created Default Collection\"\n",
    "\n",
    "\n",
    "\n",
    "def collection_lists():\n",
    "    \"\"\"\n",
    "    List All Collections available in ChromaDB\n",
    "    \"\"\"\n",
    "    collection_list = []\n",
    "    chroma_collections = chroma.list_collections()\n",
    "    if chroma_collections == None:\n",
    "        create_default_collection()\n",
    "    else:\n",
    "        for collection in chroma_collections:\n",
    "            collection_list.append(collection.name)\n",
    "    return collection_list\n",
    "\n",
    "collection_list = collection_lists()\n",
    "\n",
    "def embed_documents(collection):\n",
    "    \"\"\"\n",
    "    Given a collection name, this function loads PDF documents from a specified directory, preprocesses their content, \n",
    "    and then embeds the documents into a vector database using Chroma. \n",
    "    After embedding, it deletes the processed PDFs from the directory.\n",
    "    \n",
    "    Args:\n",
    "    - collection (str): Name of the collection to be used in the Chroma vector database.\n",
    "\n",
    "    Returns:\n",
    "    - output (str): Message indicating which documents have been embedded.\n",
    "    \"\"\"\n",
    "    loader = DirectoryLoader(\"/home/cdsw/data/\",\n",
    "                         glob=\"**/*.pdf\",\n",
    "                         loader_cls=PyPDFLoader,\n",
    "                         use_multithreading=True)\n",
    "\n",
    "    documents = loader.load()\n",
    "\n",
    "    for i in range(len(documents)):\n",
    "        documents[i].page_content = documents[i].page_content.replace('\\t', ' ')\\\n",
    "                                                         .replace('\\n', ' ')\\\n",
    "                                                         .replace('       ', ' ')\\\n",
    "                                                         .replace('      ', ' ')\\\n",
    "                                                         .replace('     ', ' ')\\\n",
    "                                                         .replace('    ', ' ')\\\n",
    "                                                         .replace('   ', ' ')\\\n",
    "                                                         .replace('  ', ' ')\n",
    "\n",
    "\n",
    "    langchain_chroma = Chroma(\n",
    "    client=chroma,\n",
    "    collection_name=collection,\n",
    "    embedding_function=embedding_function)    \n",
    "\n",
    "    collection = chroma.get_collection(collection) # Needs to be initialized as LangChain cannot add texts\n",
    "\n",
    "    # Document is chunked per page. Each page will be an entry in the Vector DB\n",
    "    for doc in documents: \n",
    "        collection.add(\n",
    "            ids=[str(uuid.uuid1())], metadatas=doc.metadata, documents=doc.page_content\n",
    "        )\n",
    "\n",
    "    pattern = \"/home/cdsw/data/*.pdf\"\n",
    "    files = glob.glob(pattern)\n",
    "    \n",
    "    output = f\"Documents have been embedded: {files}\"\n",
    "    print(output)\n",
    "    # Deleting Files\n",
    "    for file in files:\n",
    "        os.remove(file)\n",
    "    \n",
    "    return output\n",
    "\n",
    "\n",
    "##### Experimentatal Code ##### \n",
    "def set_retriver(collection_name):\n",
    "    langchain_chroma = Chroma(\n",
    "    client=chroma,\n",
    "    collection_name= collection_name,\n",
    "    embedding_function=embedding_function)\n",
    "    \n",
    "    retriever = langchain_chroma.as_retriever(search_kwargs={\"k\": 2, \"search_type\" : \"similarity\"})\n",
    "    return retriever\n",
    "\n",
    "\n",
    "# Prompt Template for Langchain\n",
    "template = \"\"\"You are a helpful AI assistant. Use only the below provided Context to answer the following question. If you do not know the answer respond with \"I don't know.\"\n",
    "Context:{context}\n",
    ">>QUESTION<<{question}\n",
    ">>ANSWER<<\"\"\"\n",
    "QA_CHAIN_PROMPT = PromptTemplate.from_template(template)\n",
    "\n",
    "  \n",
    "\n",
    "def chain(query, retriever):\n",
    "    \"\"\"\n",
    "    Executes a retrieval-based question-answering chain with specified query and retriever.\n",
    "\n",
    "    Args:\n",
    "    - query (str): The query/question to be answered.\n",
    "    - retriever (Retriever): The retriever object responsible for fetching relevant documents.\n",
    "\n",
    "    Returns:\n",
    "    - dict: Response from the RetrievalQA.\n",
    "    \"\"\"\n",
    "    qa_chain = RetrievalQA.from_chain_type(llm=llm, \n",
    "                                       chain_type=\"stuff\", \n",
    "                                       retriever=set_retriver(retriever), \n",
    "                                       return_source_documents=True,\n",
    "                                       chain_type_kwargs={\"prompt\": QA_CHAIN_PROMPT},\n",
    "                                       verbose=True, tags=[\"zephyr-7b\", \"4-bit\"]\n",
    "                                       )\n",
    "    return qa_chain(query)\n",
    "\n",
    "def add_text(history, text):\n",
    "    \"\"\"\n",
    "    Adds the user's text input to the conversation history.\n",
    "\n",
    "    Args:\n",
    "    - history (list): The existing conversation history.\n",
    "    - text (str): The user's input text.\n",
    "\n",
    "    Returns:\n",
    "    - list: Updated history with the user's input.\n",
    "    - str: Empty string (reserved for future use).\n",
    "    \"\"\"\n",
    "    history = history + [(text, None)]\n",
    "    return history, \"\"\n",
    "\n",
    "def bot(history, collection):\n",
    "    \"\"\"\n",
    "    Generates a response using a Language Model and updates the conversation history.\n",
    "\n",
    "    Args:\n",
    "    - history (list): The existing conversation history.\n",
    "    - collection (str): The name of the collection used for document retrieval.\n",
    "\n",
    "    Returns:\n",
    "    - list: Updated conversation history including the bot's response.\n",
    "    \"\"\"\n",
    "    response = llm_ans(history[-1][0], collection)\n",
    "    history[-1][1] = response\n",
    "    return history\n",
    "\n",
    "def wrap_text_preserve_newlines(text, width=110):\n",
    "    \"\"\"\n",
    "    Wraps the text while preserving newlines to fit within a specified width.\n",
    "\n",
    "    Args:\n",
    "    - text (str): The input text.\n",
    "    - width (int): The maximum width of the text.\n",
    "\n",
    "    Returns:\n",
    "    - str: Wrapped text with newlines preserved.\n",
    "    \"\"\"\n",
    "    # Split the input text into lines based on newline characters\n",
    "    lines = text.split('\\n')\n",
    "\n",
    "    # Wrap each line individually\n",
    "    wrapped_lines = [textwrap.fill(line, width=width) for line in lines]\n",
    "\n",
    "    # Join the wrapped lines back together using newline characters\n",
    "    wrapped_text = '\\n'.join(wrapped_lines)\n",
    "\n",
    "    return wrapped_text\n",
    "\n",
    "def process_llm_response(llm_response):\n",
    "    \"\"\"\n",
    "    Processes the Language Model's response by wrapping the text and printing the source documents.\n",
    "\n",
    "    Args:\n",
    "    - llm_response (dict): The response from the Language Model.\n",
    "\n",
    "    Returns:\n",
    "    - str: The wrapped text.\n",
    "    \"\"\"\n",
    "    result = wrap_text_preserve_newlines(llm_response['result'])\n",
    "    print('\\n\\nSources:')\n",
    "    for source in llm_response[\"source_documents\"]:\n",
    "        print(source.metadata['source'])\n",
    "    return result    \n",
    "\n",
    " \n",
    "\n",
    "def llm_ans(query, collection):\n",
    "    \"\"\"\n",
    "    Gets the answer from the Language Model including relevant source files.\n",
    "\n",
    "    Args:\n",
    "    - query (str): The query/question to be answered.\n",
    "    - collection (str): The name of the collection used for document retrieval.\n",
    "\n",
    "    Returns:\n",
    "    - str: The answer along with relevant source files.\n",
    "    \"\"\"\n",
    "    start = time.time()\n",
    "    llm_response = chain(query, collection)\n",
    "    end = time.time()\n",
    "    elapsed_time = end - start\n",
    "\n",
    "    \n",
    "    # print(llm_response['result'])\n",
    "    sources = []\n",
    "    for source in llm_response[\"source_documents\"]:\n",
    "        source_file = source.metadata['source']\n",
    "        source_file = source_file.replace(\"/home/cdsw/data/\", \"\")\n",
    "        sources.append(source_file)\n",
    "    source_files = \"\\n\".join(sources) \n",
    "    ans = llm_response['result'] + \"\\n \\n Relevant Sources: \\n\" + source_files + \"\\n \\n Elapsed Time: \" + str(round(elapsed_time,2)) + \" seconds\"\n",
    "    return ans\n",
    "\n",
    "def reset_state():\n",
    "    \"\"\"\n",
    "    Resets the Gradio UI\n",
    "    \"\"\"\n",
    "    return [], [], None\n",
    "\n",
    "#Gradio UI Code Block\n",
    "\n",
    "with gr.Blocks() as demo:\n",
    "    with gr.Tab(\"FileGPT\"):\n",
    "        chatbot = gr.Chatbot([], elem_id=\"chatbot\").style(height=650)\n",
    "        with gr.Row():\n",
    "            with gr.Column(scale=4):\n",
    "                with gr.Column(scale=12):\n",
    "                    user_input = gr.Textbox(show_label=False, placeholder=\"Input...\", lines=10).style(\n",
    "                        container=False)\n",
    "                with gr.Column(min_width=32, scale=1):\n",
    "                    submitBtn = gr.Button(\"Submit\", variant=\"primary\")\n",
    "            with gr.Column(scale=1):\n",
    "                collection_dropdown = gr.Dropdown(\n",
    "                    collection_list, label=\"Chroma Collections\", info=\"Choose a Collection to Query\",\n",
    "                    value = \"default\", max_choices=1\n",
    "                )\n",
    "                emptyBtn = gr.Button(\"Clear History\")\n",
    "        user_input.submit(add_text, [chatbot, user_input], [chatbot, user_input]).then(bot, [chatbot, collection_dropdown], chatbot)\n",
    "        submitBtn.click(add_text, [chatbot, user_input], [chatbot, user_input]).then(bot, [chatbot, collection_dropdown], chatbot)\n",
    "        history = gr.State([])\n",
    "        past_key_values = gr.State(None)\n",
    "        emptyBtn.click(reset_state, outputs=[chatbot, history, past_key_values], show_progress=True)\n",
    "        \n",
    "\n",
    "    with gr.Tab(\"Upload File\"):\n",
    "        with gr.Row():\n",
    "            title=\"Falcon 40B\",\n",
    "            with gr.Column(scale=4):\n",
    "                file_output = gr.File()\n",
    "                upload_button = gr.UploadButton(\"Click to Upload a File\", file_types=[\".pdf\",\".csv\",\".doc\"], file_count=\"multiple\")\n",
    "                upload_button.upload(upload_file, upload_button, file_output)\n",
    "            with gr.Column(scale=1):\n",
    "                embed_dropdown = gr.Dropdown(\n",
    "                    collection_list, label=\"Chroma Collections\", info=\"Choose a Collection to Query\", \n",
    "                    value = \"default\", max_choices=1\n",
    "                )\n",
    "                embed_button = gr.Button(\"Embed Document\", variant=\"primary\")\n",
    "                txt_3 = gr.Textbox(value=\"\", label=\"Output\")\n",
    "                \n",
    "                \n",
    "    embed_button.click(embed_documents, embed_dropdown, show_progress=True, outputs=[txt_3])\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "demo.queue()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    demo.launch(share=True,\n",
    "                enable_queue=True,\n",
    "                show_error=True,\n",
    "                server_name='127.0.0.1',\n",
    "                server_port=int(\"8091\")) \n",
    "\n",
    "    print(\"Gradio app ready\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "239dfd06-72a2-444e-954d-14e1e10e010d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
